### PR TITLE
Fix compile warnings

### PR DESCRIPTION
**Pull request by ShineQi**
get rid of some annoying warnings

### CODE REVIEW

File: src/crypto/sha512.cpp [hunk 1]
```diff
@@ -92,7 +92,8 @@ namespace fc {
    void from_variant( const variant& v, sha512& bi, uint32_t max_depth )
    {
       std::vector<char> ve = v.as< std::vector<char> >( max_depth );
-      memset( &bi, char(0), sizeof(bi) );
+      //memset( &bi, char(0), sizeof(bi) );
+      bi = sha512();
```
Comment (pmconrad) [inline src/crypto/sha512.cpp]: This is inefficient. Please add an ``inline clear() { memset... }`` method to sha512 and use that. Same for the other hash classes.

File: include/fc/array.hpp [hunk 2]
```diff
@@ -113,9 +113,11 @@ namespace fc {
     if( ve.size() )
     {
         memcpy(&bi, ve.data(), fc::min<size_t>(ve.size(),sizeof(bi)) );
+    }else{
+        for(size_t i=0;i<N;i++){
+            bi.data[i]=T();
```
Comment (pmconrad) [inline include/fc/array.hpp]: This is a functional change, zeroing out something is not necessarily the same as invoking its no-arg constructor.
If the compile complains about memset'ing a class but not memcpy'ing it, memset an array of appropriate size and memcpy that. Still less efficient than the original though, anyone got a better idea?

**Review by pmconrad — CHANGES_REQUESTED** [review]

File: include/fc/array.hpp (hunk 2 above)
Comment (ShineQi) [reply include/fc/array.hpp]: Zero initialization may not always equals setting memory to zero, 
and [here](https://en.cppreference.com/w/cpp/language/zero_initialization) is a reference

Comment (pmconrad) [reply include/fc/array.hpp]: Yes, that's what I said.

**Pull request closed without merging**